lib/source-code/source-code.js

```diff
@@ -639,6 +646,67 @@ class SourceCode extends TokenStore {
         return this.scopeManager.scopes[0];
     }
 
+    /**
+     * Gets all of the declared variables in the scope associated
+     * with `node`. This is a convenience method that passes through
+     * to the same method on the `scopeManager`.
+     * @param {ASTNode} node The node from which to retrieve the scope to check.
+     * @returns {Array<Variable>} An array of variable nodes representing
+     *      the declared variables in the scope associated with `node`.
+     */
+    getDeclaredVariables(node) {
+        return this.scopeManager.getDeclaredVariables(node);
+    }
+
+    /**
+     * Gets all the ancestors of a given node
+     * @param {ASTNode} node The node
+     * @returns {Array<ASTNode>} All the ancestor nodes in the AST, not including the provided node, starting
+     * from the root node at index 0 and going inwards to the parent node.
+     * @throws {TypeError} When `node` is missing.
+     */
+    getAncestors(node) {
+
+        if (!node) {
+            throw new TypeError("Missing required argument: node.");
+        }
+
+        const cache = this[caches].get("ancestors");
+        const cachedAncestry = cache.get(node);
+
+        /*
+         * Important! The array returned from this method must be mutable
+         * without side effects, meaning each rule must receive its own copy
+         * of the array that it can mutate without affecting other rules.
+         * Historically, this method calculated and returned a new array for
+         * each call, but that's pretty inefficient. So we cache the result
+         * and then create a new array with the same elements here.
+         */
+        if (cachedAncestry) {
+            return cachedAncestry.concat();
+        }
+
+        const ancestorsStartingAtParent = [];
+
+        for (let ancestor = node.parent; ancestor; ancestor = ancestor.parent) {
+
+            ancestorsStartingAtParent.unshift(ancestor);
```

Comment:
Huh, that is interesting. Well, if I revert to push + reverse the caching logic becomes a lot more complicated, so I'll just revert to the original form of the method and not worry about caching.

Reply:
> so I'll just revert to the original form of the method and not worry about caching

Agreed. This caching was also of somewhat questionable value as it clones arrays when storing and retrieving values from the cache, so it initially considerably degrades performance and then needs a high % of cache hits to actually have a positive effect.

Reply:
I was trying to avoid traversing up the same tree multiple times, but it appears that's not a huge time sink, so probably not worth it unless we run into a problem in the future.